internal/server/handler.go

```diff
@@ -141,3 +150,31 @@ func (s *Server) writeError(w http.ResponseWriter, sourceErr error) {
 		s.logger.Warn("failed to write error response", mlog.Err(err))
 	}
 }
+
+func (s *Server) validateRequestMatchesInstallationID(r *http.Request, installationID string) error {
+	addr := strings.Split(r.RemoteAddr, ":")[0]
+	names, err := s.lookupAddrFn(addr)
+	if err != nil {
+		return errors.Wrap(err, "failed to perform reverse domain name lookup")
+	}
+	if len(names) == 0 {
+		return errors.New("no names returned in reverse lookup")
+	}
+	name := names[0]
+
+	// Perform validation by comparing reverse lookup to expected namespace.
+	// Example Reverse Lookup:
+	//   IP_ADDR.SERVICE_NAME.NAMESPACE/INSTALLATION_ID.svc.cluster.local.
+	if !requestIsValid(name, installationID) {
+		s.logger.Warn("reverse name lookup validation failed", mlog.String("name", name), mlog.String("installationID", installationID))
```

Comment:
Let's avoid logging here because `writeError` is anyways going to log again. Rather, we can add the name and installation ID in the error.